### PR TITLE
Extended with_actions method. Now it can take array of modules.

### PR DESCRIPTION
**Pull request by WorstOfAny**

### CODE REVIEW

**Comment by AlexWayfer** [comment]
Tests, please.

**Comment by AlexWayfer** [comment]
Why not:

```ruby
class MyController < Flame::Controller
  include ImportantModule.with_actions
end
```

?

**Comment by WorstOfAny** [comment]
> Why not:

Because we would had to add this action to every module that we will use

**Comment by AlexWayfer** [comment]
> Because we would had to add this action to every module that we will use

Maybe we should write

```ruby
module ImportantModule
  extend ActionsIncluding
end
```

and just

```ruby
class MyController < Flame::Controller
  include ImprotantModule
end
```

?

**Comment by WorstOfAny** [comment]
> Maybe we should write

extend every module?
it's done inside with_actions.
so it's not obviously , for what we include modules.
more obvious seems method with_actions that tell us how module to used

**Comment by AlexWayfer** [comment]
The problem is solved in another way: the controller inherits from the class with actions, the module includes private and protected methods.

**Comment by AlexWayfer** [comment]
Or you can try this approach:

```ruby
class MyController < Flame::Controller
  include with_actions SomeModule
end
```

**Comment by WorstOfAny** [comment]
> Or you can try this approach

I choose this one

File: spec/spec_helper.rb [hunk 1]
```diff
@@ -26,3 +26,7 @@ def equal_routes(*attrs_collection)
 	routes = attrs_collection.map! { |attrs| Flame::Router::Route.new(*attrs) }
 	->(array) { array.sort == routes.sort }
 end
+
+def include_actions(mixin_actions)
+	->(obj) { (obj & mixin_actions) == mixin_actions }
```
Comment (WorstOfAny) [inline spec/spec_helper.rb]: Why iteration is better?

File: lib/flame/controller.rb [hunk 2]
```diff
@@ -156,11 +156,27 @@ def default_path
...
+			##     ....
+			##   end
+			def with_actions(mod = nil)
+				return mod.extend(ModuleMixIn) if mod
 				@with_actions ||= Class.new(self) { extend ParentActions }
 			end
 		end
 
+		## Extension to include modules for re-defining methods controller methods
+		module ModuleMixIn
```
Comment (WorstOfAny) [inline lib/flame/controller.rb]: Maybe `ExtensionForMixIn` is better?

File: spec/unit/controller_spec.rb [hunk 3]
```diff
@@ -284,5 +303,25 @@ class ControllerApplication < Flame::Application
 			inherited_controller = Class.new(ControllerController.with_actions)
 			inherited_controller.actions.should.equal ControllerController.actions
 		end
+
+		it 'should include actions from mixin module' do
+			controller_with_mixin = ControllerController.include(
+				ControllerController.with_actions(MixIn)
+			)
+			controller_with_mixin.actions.should.be include_actions(
+				MixIn.public_instance_methods
```
Comment (WorstOfAny) [inline spec/unit/controller_spec.rb]: Explain, please, why it will be better?

Comment (AlexWayfer) [inline spec/unit/controller_spec.rb]: In tests it is better to duplicate the information explicitly. If they then start to fall after the changes - there's nothing wrong with that, on the contrary, it's clearer what has changed, and whether it is necessary to change something.

File: lib/flame/controller.rb (hunk 2 above)
Comment (AlexWayfer) [inline lib/flame/controller.rb]: Ruby has no `MixIn`s, Ruby has modules…

File: spec/spec_helper.rb (hunk 1 above)
Comment (AlexWayfer) [inline spec/spec_helper.rb]: No new objects (`Array`), no repetitions…

File: lib/flame/controller.rb (hunk 2 above)
Comment (WorstOfAny) [inline lib/flame/controller.rb]: `ExtensionForMixInModule` then

Comment (AlexWayfer) [inline lib/flame/controller.rb]: Every module is namespace or extension…

So, it's an unnecessary clarification.

And why `MixInModule`? They are synonyms.

Comment (WorstOfAny) [inline lib/flame/controller.rb]: because this module extends module from `with_actions` arguments. It isn't Module with actions.

Comment (AlexWayfer) [inline lib/flame/controller.rb]: Extended module is module with actions.

Even before it's extended.

Okay, another one: `IncludingActions`.

Comment (WorstOfAny) [inline lib/flame/controller.rb]: > Even before it's extended.

but we talk about module that extends Module with actions.

> Okay, another one: IncludingActions.

OK.

File: spec/unit/controller_spec.rb [hunk 4]
```diff
@@ -284,5 +293,12 @@ class ControllerApplication < Flame::Application
 			inherited_controller = Class.new(ControllerController.with_actions)
 			inherited_controller.actions.should.equal ControllerController.actions
 		end
+
+		it 'should include actions from mixin module' do
+			controller_with_mixin = ControllerController.with_actions(MixIn)
+			controller_with_mixin.actions.should.be include_actions(
+				MixIn.instance_methods + MixIn.private_instance_methods
```
Comment (WorstOfAny) [inline spec/unit/controller_spec.rb]: was done.

File: spec/unit/controller_spec.rb [hunk 5]
```diff
@@ -284,5 +303,25 @@ class ControllerApplication < Flame::Application
 			inherited_controller = Class.new(ControllerController.with_actions)
 			inherited_controller.actions.should.equal ControllerController.actions
 		end
+
+		it 'should include actions from mixin module' do
+			controller_with_mixin = ControllerController.include(
+				ControllerController.with_actions(MixIn)
+			)
```
Comment (WorstOfAny) [inline spec/unit/controller_spec.rb]: Done

File: spec/unit/controller_spec.rb (hunk 3 above)
Comment (WorstOfAny) [inline spec/unit/controller_spec.rb]: Okey. Done

File: spec/unit/controller_spec.rb [hunk 6]
```diff
@@ -284,5 +303,25 @@ class ControllerApplication < Flame::Application
...
+			controller_with_mixin.actions.should.be include_actions(
+				MixIn.public_instance_methods
+			)
+		end
+
+		it 'should include actions from several mixin module' do
```
Comment (WorstOfAny) [inline spec/unit/controller_spec.rb]: Removed.

File: lib/flame/controller.rb [hunk 7]
```diff
@@ -156,11 +156,27 @@ def default_path
 			## @example Inherit controller with parent actions by method
 			##   class MyController < BaseController.with_actions
 			##   end
-			def with_actions
+			## @example Include module actions to controller
```
Comment (WorstOfAny) [inline lib/flame/controller.rb]: Changed.

File: spec/spec_helper.rb (hunk 1 above)
Comment (WorstOfAny) [inline spec/spec_helper.rb]: Changed.

File: spec/unit/controller_spec.rb [hunk 8]
```diff
@@ -284,5 +303,25 @@ class ControllerApplication < Flame::Application
 			inherited_controller = Class.new(ControllerController.with_actions)
 			inherited_controller.actions.should.equal ControllerController.actions
 		end
+
+		it 'should include actions from mixin module' do
+			controller_with_mixin = ControllerController.include(
+				ControllerController.with_actions(MixIn)
+			)
+			controller_with_mixin.actions.should.be include_actions(
```
Comment (WorstOfAny) [inline spec/unit/controller_spec.rb]: Removed.

File: lib/flame/controller.rb [hunk 9]
```diff
@@ -177,11 +177,27 @@ def default_path
...
 		end
 
+		## Extension to include modules for re-defining methods controller methods
+		module IncludingActions
+			def included(ctrl)
+				public_instance_methods.each do |meth|
+					ctrl.send :define_method, meth, public_instance_method(meth)
+				end
+			end
+		end
```
Comment (AlexWayfer) [inline lib/flame/controller.rb]: This module is very similar to `ParentActions`…

File: spec/unit/controller_spec.rb [hunk 10]
```diff
@@ -68,6 +68,17 @@ class IndexController < Flame::Controller
 	end
 end
 
+## Module for Controller tests
+module MixIn
```
Comment (AlexWayfer) [inline spec/unit/controller_spec.rb]: `SomeActions` maybe?

File: spec/unit/controller_spec.rb [hunk 11]
```diff
@@ -365,5 +376,12 @@ class ControllerApplication < Flame::Application
 			inherited_controller = Class.new(ControllerController.with_actions)
 			inherited_controller.actions.should.equal ControllerController.actions
 		end
+
+		it 'should include actions from mixin module' do
+			controller_with_mixin = ControllerController
+				.class_exec { include with_actions MixIn }
+			controller_with_mixin.actions
+				.should include_actions %i[mixed another_mixed]
```
Comment (AlexWayfer) [inline spec/unit/controller_spec.rb]: You can add `should.not include_actions` for private methods.

File: lib/flame/controller.rb (hunk 9 above)
Comment (WorstOfAny) [inline lib/flame/controller.rb]: Similar, but not the same. You can't do it the same way

File: spec/unit/controller_spec.rb (hunk 11 above)
Comment (WorstOfAny) [inline spec/unit/controller_spec.rb]: I can add `not_include_actions`, we can't use the same method. Or i can add another argument to this method, that represents `:none?` and `:all?` methods.

File: lib/flame/controller.rb (hunk 9 above)
Comment (AlexWayfer) [inline lib/flame/controller.rb]: Maybe check what was extended (`class` or `module`) in hook can help you?

Comment (WorstOfAny) [inline lib/flame/controller.rb]: something like this :
```ruby
def define_actions
				if is_a? Class
					superclass.actions.each do |public_method|
						um = superclass.public_instance_method(public_method)
						define_method public_method, um
					end
				else
					define_singleton_method :included do |ctrl|
						public_instance_methods.each do |meth|
							ctrl.send :define_method, meth, public_instance_method(meth)
						end
					end
				end
			end
```
or : 
```ruby
			def self.extended(ctrl)
				ctrl.public_send(
					ctrl.is_a?(Class) ? :define_parent_actions : :define_module_actions
				)
			end
```
?

Comment (AlexWayfer) [inline lib/flame/controller.rb]: `.public_instance_method(public_method)` is repeated. Just `superclass` or `self`, I guess.

`define_method public_method, um` is also repeated. Just `self` or `ctrl`, I guess.

Comment (WorstOfAny) [inline lib/flame/controller.rb]: module haven't method `actions`.

Comment (AlexWayfer) [inline lib/flame/controller.rb]: I know. And what? :)

Comment (WorstOfAny) [inline lib/flame/controller.rb]: i can't embody what you want.

Comment (WorstOfAny) [inline lib/flame/controller.rb]: Changed.

File: spec/unit/controller_spec.rb (hunk 10 above)
Comment (WorstOfAny) [inline spec/unit/controller_spec.rb]: Changed.

File: spec/unit/controller_spec.rb (hunk 11 above)
Comment (WorstOfAny) [inline spec/unit/controller_spec.rb]: Changed.

File: lib/flame/controller.rb (hunk 9 above)
Comment (WorstOfAny) [inline lib/flame/controller.rb]: Returned to first solution

File: lib/flame/controller.rb [hunk 12]
```diff
@@ -178,11 +178,28 @@ def default_path
...
+			##     ....
+			##   end
+			def with_actions(mod = nil)
+				return mod.extend(ModuleActions) if mod
 				@with_actions ||= Class.new(self) { extend ParentActions }
 			end
 		end
 
+		## Extension for modules whose public methods will be defined as actions
+		##   via including
```
Comment (AlexWayfer) [inline lib/flame/controller.rb]: [Spaces are unnecessary](http://www.rubydoc.info/gems/yard/file/docs/GettingStarted.md).

Comment (WorstOfAny) [inline lib/flame/controller.rb]: Changed

**Pull request merged**